share/vss.go

```diff
@@ -0,0 +1,493 @@
+		return nil, nil, err
+	} else if !bytes.Equal(sid, d.SessionID) {
+		return nil, nil, errors.New("verifier: invalid session id in Deal")
+	}
+
+	if v.aggregator == nil {
+		v.aggregator = newAggregator(v.suite, v.verifiers, d.Commitments, t, d.SessionID)
+	}
+
+	sig, err := sign.Schnorr(v.suite, v.long, sid)
+	if err := v.verifyDeal(d, true); err != nil {
+		if err == errDealAlreadyReceived {
+			return nil, nil, err
+		}
+		complaint := &Complaint{
+			SessionID: sid,
+			Public:    v.pub,
+			Deal:      d,
+			Signature: sig,
+		}
+		// add it to the list of complaints directly, no need to verify
+		v.complaints[v.pub.String()] = complaint
+		return nil, complaint, err
+	}
+	approval := &Approval{
+		Public:    v.pub,
+		SessionID: v.sid,
+		Signature: sig,
+	}
+	return approval, nil, nil
+}
+
+// ReceiveComplaints takes a Complaint and stores it if the Complaint is valid.
+// If the verifier already saw a Complaint from the same origin, it returns an error.
+// If the Complaint is not valid, it returns an error. That does NOT mean the
+// Deal is not good, but rather only this Complaint is not valid and should not
+// be treated.  To know whether the deal is good, call first v.EnoughApproval(),
+// and if true, call v.IsDealGood().
+func (v *Verifier) ReceiveComplaint(c *Complaint) error {
+	return v.verifyComplaint(c)
+}
+
+// ReceiveDealerResponse takes a DealerResponse and returns an error if
+// something went wrong during the verification. If it is the case, that
+// probably means the Dealer is acting maliciously. In order to be sure, call
+// v.EnoughApprovals() and if true, v.IsDealGood().
+// NOTE: it does *not* verify the complaint associated with since it is supposed
+// to already have received it.
+func (v *Verifier) ReceiveDealerResponse(dr *DealerResponse) error {
+	pub := dr.Complaint.Public
+	if _, ok := v.aggregator.complaints[pub.String()]; !ok {
+		return errors.New("verifier: no complaints received for this response")
+	}
+	// XXX should we verify the corectness of the complaint and make sure we
+	// actually received it ?
+
+	if err := v.verifyDeal(dr.Deal, false); err != nil {
+		// if one response is bad, flag the dealer as malicious
+		v.dealerBad = true
+		return err
+	}
+	return nil
+}
+
+func (v *Verifier) ReceiveApproval(ap *Approval) error {
+	return v.verifyApproval(ap)
+}
+
+func (v *Verifier) DealCertified() bool {
+	return v.aggregator.DealCertified() && !v.dealerBad
+}
+
+// Complaint is a message that must be broadcasted to every verifiers when
+// a verifier receives an invalid Deal.
+type Complaint struct {
+	Public    abstract.Point
+	SessionID []byte
+	Deal      *Deal
+	// Signature over the msg
+	// H(Index || deal.MarshalBinary() || verifiers)
+	Signature []byte
+}
+
+// DealerResponse is a message that is broadcasted by the Dealer in response to
+// a Complaint. It contains the original Complaint as well as the shares
+// distributed to the complainer.
+type DealerResponse struct {
```

Comment:
Hmmm I wonder if this should be renamed to `ResponseComplaint`, `ComplaintResponse` or even just `Response` since the `Complaint` struct also does not contain a `Verifier` label?!